File: app/src/test/java/com/babestudios/companyinfouk/ui/search/SearchPresenterTest.java
package com.babestudios.companyinfouk.ui.search;

import com.babestudios.companyinfouk.data.DataManager;
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult;
import com.babestudios.companyinfouk.data.model.search.SearchHistoryItem;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;

import io.reactivex.Observable;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class SearchPresenterTest {

	private SearchPresenter searchPresenter;

	private SearchActivityView view;

	private SearchHistoryItem searchHistoryItem = new SearchHistoryItem("TUI", "12344", 12L);

	@Before
	public void setUp() {
		searchPresenter = new SearchPresenter(Mockito.mock(DataManager.class));
		searchPresenter.create();
		view = mock(SearchActivityView.class);
		searchPresenter.attachView(view);
		when(searchPresenter.dataManager.searchCompanies(any(), anyString())).thenReturn(Observable.just(new CompanySearchResult()));
		ArrayList<SearchHistoryItem> searchHistoryItems = new ArrayList<>();
		SearchHistoryItem searchHistoryItem1 = new SearchHistoryItem("RUN", "12345", 12L);
		searchHistoryItems.add(searchHistoryItem);
		searchHistoryItems.add(searchHistoryItem1);
		when(searchPresenter.dataManager.addRecentSearchItem(searchHistoryItem)).thenReturn(searchHistoryItems);
	}

	@Test
	public void whenFabClickedInStateRecentSearches_thenShowDeleteRecentSearchesDialogIsCalled() {
		searchPresenter.onFabClicked();
		verify(view).showDeleteRecentSearchesDialog();
	}

	@Test
	public void whenFabClickedInStateSearch_thenClearSearchView_andRefreshAdapter_andShowRecentSearchesCalled() {
		searchPresenter.search("");
		searchPresenter.onFabClicked();
		verify(view).clearSearchView();
		verify(view).refreshRecentSearchesAdapter(any());
		verify(view, times(2)).showRecentSearches(any());
		verify(view, times(2)).changeFabImage(SearchPresenter.FabImage.FAB_IMAGE_RECENT_SEARCH_DELETE);
	}

	@Test
	public void whenGetCompany_thenDataManagerAddRecentSearchItem_andStartActivityIsCalled() {
		searchPresenter.getCompany("TUI", "12344");
		verify(view).startCompanyActivity(anyString(), anyString());
		verify(searchPresenter.dataManager).addRecentSearchItem(searchHistoryItem);
	}

	@Test
	public void whenSearch_thenDataManagerSearchCompaniesIsCalled() {
		searchPresenter.search("");
		verify(searchPresenter.dataManager).searchCompanies(any(), anyString());
	}

	@Test
	public void whenSearchLoadMore_thenDataManagerSearchCompaniesIsCalled() {
		searchPresenter.search("");
		verify(searchPresenter.dataManager).searchCompanies(any(), anyString());
	}


}

File: app/src/test/java/com/babestudios/companyinfouk/ui/search/SearchPresenterTest.kt
package com.babestudios.companyinfouk.ui.search

import com.babestudios.companyinfouk.CompaniesHouseApplication
import com.babestudios.companyinfouk.DaggerTestApplicationComponent
import com.babestudios.companyinfouk.TestApplicationModule
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult
import com.babestudios.companyinfouk.data.model.search.SearchHistoryItem
import com.babestudios.companyinfouk.ui.main.ContentChange
import com.babestudios.companyinfouk.ui.main.MainPresenter
import com.babestudios.companyinfouk.ui.main.MainViewModel
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.whenever
import io.reactivex.CompletableSource
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class SearchPresenterTest {

	private lateinit var mainPresenter: MainPresenter

	private val searchHistoryItem = SearchHistoryItem("TUI", "12344", 12L)

	@Before
	fun setUp() {
		val testApplicationComponent = DaggerTestApplicationComponent.builder()
				.testApplicationModule(TestApplicationModule(CompaniesHouseApplication()))
				.build()
		mainPresenter = testApplicationComponent.mainPresenter()
		//whenever(officersPresenter.companiesRepository.getOfficers("0", "0")).thenReturn(Single.just(Officers()))
		val officersViewModel = MainViewModel()
		//officersViewModel.state.value.companyNumber = "0"
		whenever(mainPresenter.companiesRepository.searchCompanies(any(), anyString())).thenReturn(Single.just(CompanySearchResult()))
		val searchHistoryItems = ArrayList<SearchHistoryItem>()
		val searchHistoryItem1 = SearchHistoryItem("TUI", "12345", 12L)
		searchHistoryItems.add(searchHistoryItem)
		searchHistoryItems.add(searchHistoryItem1)
		whenever(mainPresenter.companiesRepository.addRecentSearchItem(searchHistoryItem)).thenReturn(searchHistoryItems)
		mainPresenter.setViewModel(officersViewModel, CompletableSource { })
	}

	@Test
	fun whenFabClickedInStateRecentSearches_thenShowDeleteRecentSearchesDialogIsCalled() {
		mainPresenter.fabMainClicked()
		assert(mainPresenter.viewModel.state.value.contentChange == ContentChange.DELETE_SEARCH_HISTORY)
	}

	@Test
	fun `when searchItemClicked then addRecentSearchItem and StartActivity Is Called`() {
		mainPresenter.searchItemClicked("TUI", "12344")
		//TODO This uses context in Presenter. Create a string provider to get this from strings.xml, but don't rely on context here
		//assert(mainPresenter.viewModel.state.value.contentChange == ContentChange.SEARCH_HISTORY_ITEMS_UPDATED)
		verify(mainPresenter.companiesRepository).addRecentSearchItem(searchHistoryItem)
	}

	@Test
	fun whenSearch_thenDataManagerSearchCompaniesIsCalled() {
		mainPresenter.search("")
		verify(mainPresenter.companiesRepository).searchCompanies(any(), anyString())
	}

	@Test
	fun whenSearchLoadMore_thenDataManagerSearchCompaniesIsCalled() {
		mainPresenter.search("")
		verify(mainPresenter.companiesRepository).searchCompanies(any(), anyString())
	}


}
